src/main/java/com/_up/megastore/services/implementations/UserService.java

```diff
@@ -175,19 +177,37 @@ public void sendEmailToRecoverPassword(String email) {
   }
 
   private User findUserByEmailOrThrowException(String email) {
-    return userRepository.findByEmailAndDeletedIsFalse(email).orElseThrow(() -> new RuntimeException("User not found"));
+    return userRepository.findByEmailAndDeletedIsFalse(email).orElseThrow(() -> new RuntimeException("User with email " + email +" does not exists."));
   }
 
   @Override
-  public void recoverPassword(UUID userId, String newPassword, UUID recoverPasswordToken) {
-    User user = findUserToRecoverPasswordOrThrowException(userId, recoverPasswordToken);
-    user.setPassword(passwordEncoder.encode(newPassword));
+  public void recoverPassword(UUID userId, String password, String confirmPassword, UUID recoverPasswordToken) {
+    User user = findUserByIdOrThrowException(userId);
+    ifTokenIsNotTheSameThrowException(recoverPasswordToken, user);
+    ifPasswordIsNotStrongerThrowException(password);
+    ifPasswordsAreNotEqualsThrowException(password, confirmPassword);
+    user.setPassword(passwordEncoder.encode(password));
     userRepository.save(user);
   }
 
-  private User findUserToRecoverPasswordOrThrowException(UUID userId, UUID recoverPasswordToken) {
-    return userRepository.findByUserIdAndRecoverPasswordTokenIs(userId, recoverPasswordToken)
-            .orElseThrow(() -> new RuntimeException("Token is invalid or user not exists."));
+  private void ifTokenIsNotTheSameThrowException(UUID recoverPasswordToken, User user) {
+    if (!user.getRecoverPasswordToken().equals(recoverPasswordToken))
+      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "The token is not the same");
+  }
+
+  private void ifPasswordIsNotStrongerThrowException(String password) {
+    String regexStrongPassword = "^(?=.*[A-Z])(?=.*[a-z])(?=.*[0-9])(?=.*[!¡¿?$%&#@_-])[A-Za-z0-9!¡¿?$%&#@_-]+$";
+    if (!password.matches(regexStrongPassword))
+      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Password is not stronger");
+  }
+
+  private void ifPasswordsAreNotEqualsThrowException(String password, String confirmPassword) {
+    if (!password.equals(confirmPassword))
+      throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Passwords do not match");
+  }
+
+  private User findUserByIdOrThrowException(UUID userId) {
+    return userRepository.findById(userId).orElseThrow(() -> new NoSuchElementException("User with id " + userId + "does not exist"));
```

Comment:
Esta [exception](https://docs.oracle.com/javase/8/docs/api/java/util/NoSuchElementException.html) me acabo de dar cuenta que se usa en varios lados, este tipo de exception no representa a lo que realmente queremos decir cuando intentamos encontrar un usuario que no se encuentra sino que es para otra cosa, acceder a los elementos de un iterador por ejemplo.

Para esto mejor usar el ResponseStatusException

```suggestion
    return userRepository.findById(userId).orElseThrow(() -> new ResponseStatusException(HttpStatus.BAD_REQUEST, "User with id " + userId + " does not exist"));
```

Agrego un espacio porque sino va a quedar pegado el userId al "does"